File: backend/app/geometry.py
"""Pure geometry math — no DB/HTTP dependencies, fully unit-testable.

Mirrors frontend/src/canvas/geometry.ts intentionally: the client needs
zero-latency feedback while drawing, the server needs the same math to
render the PDF from persisted data. Kept in sync by the shared unit tests
(10ft x 12ft rectangle is the canonical regression case on both sides).
"""

import math


def edge_length(p1: tuple[float, float], p2: tuple[float, float]) -> float:
    """Distance between two points, in the same unit as the inputs (inches)."""
    return math.hypot(p2[0] - p1[0], p2[1] - p1[1])


def shoelace_area(vertices: list[tuple[float, float]]) -> float:
    """Area of a closed polygon in square-inches, via the shoelace formula.

    `vertices` should be the ordered polygon points (not repeating the first
    point at the end). Works for convex and simple concave polygons.
    """
    n = len(vertices)
    if n < 3:
        return 0.0
    total = 0.0
    for i in range(n):
        x1, y1 = vertices[i]
        x2, y2 = vertices[(i + 1) % n]
        total += x1 * y2 - x2 * y1
    return abs(total) / 2.0


def square_inches_to_square_feet(area_sq_in: float) -> float:
    return area_sq_in / 144.0

File: frontend/src/canvas/geometry.ts
// Mirrors backend/app/geometry.py intentionally: the client needs
// zero-latency feedback while drawing (live dimension/area labels), the
// server needs the same math to render the PDF from persisted data.
// Kept in sync via the shared 10ft x 12ft rectangle test case on both sides.

import type { Point } from '../types'

export function edgeLength(p1: Point, p2: Point): number {
  return Math.hypot(p2.x - p1.x, p2.y - p1.y)
}

export function shoelaceArea(vertices: Point[]): number {
  const n = vertices.length
  if (n < 3) return 0
  let total = 0
  for (let i = 0; i < n; i++) {
    const { x: x1, y: y1 } = vertices[i]
    const { x: x2, y: y2 } = vertices[(i + 1) % n]
    total += x1 * y2 - x2 * y1
  }
  return Math.abs(total) / 2
}

export function polygonCentroid(vertices: Point[]): Point {
  const n = vertices.length
  if (n === 0) return { x: 0, y: 0 }
  const sum = vertices.reduce((acc, v) => ({ x: acc.x + v.x, y: acc.y + v.y }), { x: 0, y: 0 })
  return { x: sum.x / n, y: sum.y / n }
}

/**
 * Returns the solid (un-gapped) segments of a wall, accounting for openings.
 * Each entry is [startInches, endInches] along the wall from its start vertex.
 * Openings are clamped to [0, wallLength] and sorted before processing.
 */
export function computeSolidSegments(
  wallLengthInches: number,
  openings: { offset_along_edge: number; width: number }[],
): [number, number][] {
  const sorted = [...openings].sort((a, b) => a.offset_along_edge - b.offset_along_edge)
  const segments: [number, number][] = []
  let pos = 0
  for (const o of sorted) {
    const oStart = Math.max(0, o.offset_along_edge)
    const oEnd = Math.min(wallLengthInches, o.offset_along_edge + o.width)
    if (oEnd <= oStart) continue
    if (oStart > pos) segments.push([pos, oStart])
    pos = oEnd
  }
  if (pos < wallLengthInches) segments.push([pos, wallLengthInches])
  return segments
}
